include/unifex/exception.hpp

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <exception>
+#include <utility>
+#include <type_traits>
+
+#include <unifex/std_concepts.hpp>
+
+#include <unifex/detail/prologue.hpp>
+
+// Microsoft's STL has a fast implementation of std::make_exception_ptr.
+// So does libc++ when targetting Microsoft's ABI. On other stdlibs,
+// move as much exception machinery into a cpp file as possible to avoid
+// template code bloat.
+#if !defined(_LIBCPP_ABI_MICROSOFT) && !defined(_MSVC_STL_VERSION)
+#define UNIFEX_HAS_FAST_MAKE_EXCEPTION_PTR() 0
```

Comment:
The precedent for naming and form of this style of "is this thing available" macros is.
```
#define UNIFEX_HAVE_FILESYSTEM 1
```

Can we change this to:
```
#define UNIFEX_HAVE_FAST_MAKE_EXCEPTION_PTR 0
```

Reply:
Also, is this macro _really_ about it being "runtime fast" or is it more about reducing binary size?

Reply:
My preference would be to change all the config macros to function-like macros like I'm doing here. [It's safer.](https://www.fluentcpp.com/2019/05/28/better-macros-better-flags/)

The issue is that sometimes we might do something like this:

```c++
#define SOME_MACRO 0
```

and then erroneously use it like

```c++
#ifdef SOME_MACRO
```

If it's a function-like macro, that misuse is immediately flagged.